File: packages/shared/src/cognitive/execution-record.ts
/**
 * ExecutionRecord — resultado de un intento de envío (ADR-034 OI-3).
 * UNKNOWN ≠ ERROR. Excepción tras intentar enviar ≠ no-ejecutado.
 * ≠ ExecutionPlan (F4) ≠ ExecuteTrade ≠ broker.
 */

export type ExecutionOutcomeV1 =
  | "not_executed"
  | "executed"
  | "error"
  | "unknown";

export type ExecutionRecordV1 = {
  outcome: ExecutionOutcomeV1;
  reason: string | null;
  transactionId: string | null;
  sendAttempted: boolean;
};

export const EXECUTION_RECORD_KEY = "executionRecord";

export type BuildExecutionRecordInputV1 = {
  filled?: boolean;
  sendAttempted?: boolean;
  transactionId?: string | null;
  exception?: string | null;
  notExecutedReason?: string | null;
};

function nonEmpty(value: string | null | undefined): string | null {
  if (typeof value !== "string") return null;
  const trimmed = value.trim();
  return trimmed ? trimmed : null;
}

/**
 * Kernel: filled → executed; send without fill → unknown;
 * pre-send exception → error; else not_executed.
 */
export function buildExecutionRecord(
  input: BuildExecutionRecordInputV1 = {},
): ExecutionRecordV1 {
  const transactionId = nonEmpty(input.transactionId ?? null);
  const exception = nonEmpty(input.exception ?? null);
  const notExecutedReason = nonEmpty(input.notExecutedReason ?? null);

  if (input.filled === true) {
    return {
      outcome: "executed",
      reason: null,
      transactionId,
      sendAttempted: true,
    };
  }
  if (input.sendAttempted === true) {
    return {
      outcome: "unknown",
      reason: exception ?? "execute_exception",
      transactionId: null,
      sendAttempted: true,
    };
  }
  if (exception) {
    return {
      outcome: "error",
      reason: exception,
      transactionId: null,
      sendAttempted: false,
    };
  }
  return {
    outcome: "not_executed",
    reason: notExecutedReason,
    transactionId: null,
    sendAttempted: false,
  };
}

/** Copy de mesa: unknown nunca se lee como «no se ejecutó». */
export function executionOutcomeCopy(outcome: ExecutionOutcomeV1): string {
  switch (outcome) {
    case "executed":
      return "Ejecutado";
    case "not_executed":
      return "No se envió";
    case "error":
      return "Error antes de enviar (no ejecutado)";
    case "unknown":
      return "Resultado desconocido — no asumir que no se ejecutó";
  }
}

File: packages/py/analytics/src/bolsa_analytics/cognitive/execution_record.py
"""ExecutionRecord — resultado de un intento de envío (ADR-034 OI-3).

UNKNOWN ≠ ERROR. Excepción tras intentar enviar ≠ no-ejecutado.
≠ ExecutionPlan (F4) ≠ ExecuteTrade ≠ broker.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

ExecutionOutcome = Literal["not_executed", "executed", "error", "unknown"]

EXECUTION_RECORD_KEY = "executionRecord"


def _non_empty(value: object) -> str | None:
    if not isinstance(value, str):
        return None
    trimmed = value.strip()
    return trimmed or None


@dataclass(frozen=True, slots=True)
class ExecutionRecord:
    """Foto honesta del intento de fill. filled gana a exception (OI-1)."""

    outcome: ExecutionOutcome
    reason: str | None
    transaction_id: str | None
    send_attempted: bool

    def to_dict(self) -> dict[str, object]:
        return {
            "outcome": self.outcome,
            "reason": self.reason,
            "transactionId": self.transaction_id,
            "sendAttempted": self.send_attempted,
        }


def build_execution_record(
    *,
    filled: bool = False,
    send_attempted: bool = False,
    transaction_id: str | None = None,
    exception: str | None = None,
    not_executed_reason: str | None = None,
) -> ExecutionRecord:
    """Kernel: filled → executed; send without fill → unknown;

    pre-send exception → error; else not_executed.
    """
    tx = _non_empty(transaction_id)
    exc = _non_empty(exception)
    skip_reason = _non_empty(not_executed_reason)

    if filled:
        return ExecutionRecord(
            outcome="executed",
            reason=None,
            transaction_id=tx,
            send_attempted=True,
        )
    if send_attempted:
        return ExecutionRecord(
            outcome="unknown",
            reason=exc or "execute_exception",
            transaction_id=None,
            send_attempted=True,
        )
    if exc:
        return ExecutionRecord(
            outcome="error",
            reason=exc,
            transaction_id=None,
            send_attempted=False,
        )
    return ExecutionRecord(
        outcome="not_executed",
        reason=skip_reason,
        transaction_id=None,
        send_attempted=False,
    )


def execution_outcome_copy(outcome: ExecutionOutcome) -> str:
    """Copy de mesa: unknown nunca se lee como «no se ejecutó»."""
    if outcome == "executed":
        return "Ejecutado"
    if outcome == "not_executed":
        return "No se envió"
    if outcome == "error":
        return "Error antes de enviar (no ejecutado)"
    return "Resultado desconocido — no asumir que no se ejecutó"
